File: src/main/java/com/adven/concordion/extensions/exam/PlaceholdersResolver.java
package com.adven.concordion.extensions.exam;

import com.adven.concordion.extensions.exam.db.Range;
import org.concordion.api.Evaluator;
import org.joda.time.*;
import org.joda.time.base.BaseSingleFieldPeriod;

import java.util.Date;

import static java.lang.Integer.parseInt;
import static org.joda.time.LocalDateTime.fromDateFields;
import static org.joda.time.LocalDateTime.now;
import static org.joda.time.format.DateTimeFormat.forPattern;

public class PlaceholdersResolver {
    private static final String PREFIX_JSON_UNIT_ALIAS = "!{";
    private static final String PREFIX_EXAM = "${exam.";
    private static final String PREFIX_VAR = "${var.";
    private static final char POSTFIX = '}';

    public static String resolveJson(String body, Evaluator eval) {
        return resolve(body, "json", eval);
    }

    public static String resolveXml(String body, Evaluator eval) {
        return resolve(body, "xml", eval);
    }

    private static String resolve(String body, String type, Evaluator eval) {
        return resolveAliases(
                type,
                resolveExamCommands(
                        resolveVars(body, eval)
                )
        );
    }

    private static String resolveVars(String body, Evaluator eval) {
        while (body.contains(PREFIX_VAR)) {
            String original = body;
            String var = extractVarFrom(original, "var");
            body = original.replace(PREFIX_VAR + var + POSTFIX, getObject(eval, var).toString());
        }
        return body;
    }

    private static Object getObject(Evaluator eval, String var) {
        Object variable = eval.getVariable("#" + var);
        return variable == null ? eval.evaluate(var.contains(".") ? "#" + var : var) : variable;
    }

    private static String resolveExamCommands(String body) {
        while (body.contains(PREFIX_EXAM)) {
            String original = body;
            String var = extractVarFrom(original, "exam");
            if (var.contains(":")) {
                String[] varAndFormat = var.split(":", 2);
                String date = forPattern(varAndFormat[1]).print(fromDateFields((Date) constants(varAndFormat[0])));
                original = original.replace(PREFIX_EXAM + var + POSTFIX, date);
            } else {
                original = original.replace(PREFIX_EXAM + var + POSTFIX, constants(var).toString());
            }
            body = original;
        }
        return body;
    }

    private static String resolveAliases(String type, String body) {
        while (body.contains(PREFIX_JSON_UNIT_ALIAS)) {
            String original = body;
            String alias = extractFromAlias(original);
            body = original.replace(PREFIX_JSON_UNIT_ALIAS + alias + POSTFIX, toPlaceholder(alias, type));
        }
        return body;
    }

    private static String toPlaceholder(String alias, String type) {
        String result;
        switch (alias.toLowerCase()) {
            case "any-string":
            case "string":
            case "str":
                result = "${" + type + "-unit.any-string}";
                break;
            case "any-number":
            case "number":
            case "num":
                result = "${" + type + "-unit.any-number}";
                break;
            case "any-boolean":
            case "boolean":
            case "bool":
                result = "${" + type + "-unit.any-boolean}";
                break;
            case "regexp":
            case "regex":
                result = "${" + type + "-unit.regex}";
                break;
            case "ignored":
            case "ignore":
                result = "${" + type + "-unit.ignore}";
                break;
            default:
                result = String.format("${" + type + "-unit.matches:%s}%s", (Object[]) alias.split(" "));
        }
        return result;
    }

    private static Object constants(String var) {
        if (var.startsWith("date(")) {
            String date = var.substring("date(".length(), var.indexOf(")"));
            return LocalDateTime.parse(date, forPattern("dd.MM.yyyy")).toDate();
        } else if (var.startsWith("now+")) {
            return now().plus(parsePeriod(var)).toDate();
        } else if (var.startsWith("now-")) {
            return now().minus(parsePeriod(var)).toDate();
        }
        switch (var) {
            case "yesterday":
                return now().minusDays(1).toDate();
            case "today":
            case "now":
                return now().toDate();
            case "tomorrow":
                return now().plusDays(1).toDate();
            default:
                return null;
        }
    }

    private static Period parsePeriod(String var) {
        Period p = Period.ZERO;
        String[] periods = var.substring(5, var.indexOf("]")).split(",");
        for (String period : periods) {
            String[] parts = period.trim().split(" ");
            p = isValue(parts[0])
                    ? p.plus(periodBy(parseInt(parts[0]), parts[1]))
                    : p.plus(periodBy(parseInt(parts[1]), parts[0]));
        }
        return p;
    }

    public static BaseSingleFieldPeriod periodBy(int val, String type) {
        switch (type) {
            case "d":
            case "day":
            case "days":
                return Days.days(val);
            case "M":
            case "month":
            case "months":
                return Months.months(val);
            case "y":
            case "year":
            case "years":
                return Years.years(val);
            case "h":
            case "hour":
            case "hours":
                return Hours.hours(val);
            case "m":
            case "min":
            case "minute":
            case "minutes":
                return Minutes.minutes(val);
            case "s":
            case "sec":
            case "second":
            case "seconds":
                return Seconds.seconds(val);
            default:
                throw new UnsupportedOperationException("Unsupported period type " + type);
        }
    }

    private static boolean isValue(String part) {
        try {
            parseInt(part);
        } catch (NumberFormatException e) {
            return false;
        }
        return true;
    }

    public static Object resolveToObj(String placeholder, Evaluator evaluator) {
        if (placeholder.contains(PREFIX_VAR)) {
            return getObject(evaluator, extractVarFrom(placeholder, "var"));
        } else if (placeholder.contains(PREFIX_EXAM)) {
            return constants(extractVarFrom(placeholder, "exam"));
        } else if (Range.isRange(placeholder)) {
            return Range.from(placeholder);
        }
        return placeholder;
    }

    private static String extractVarFrom(String placeholder, final String namespace) {
        String s = placeholder.substring(placeholder.indexOf("${" + namespace + "."));
        return s.substring(s.indexOf('.') + 1, s.indexOf(POSTFIX));
    }

    private static String extractFromAlias(String placeholder) {
        String s = placeholder.substring(placeholder.indexOf(PREFIX_JSON_UNIT_ALIAS));
        return s.substring(2, s.indexOf(POSTFIX));
    }
}

File: src/main/java/com/adven/concordion/extensions/exam/PlaceholdersResolver.kt
@file:JvmName("PlaceholdersResolver")

package com.adven.concordion.extensions.exam

import org.concordion.api.Evaluator
import org.joda.time.*
import org.joda.time.LocalDateTime.fromDateFields
import org.joda.time.LocalDateTime.now
import org.joda.time.base.BaseSingleFieldPeriod
import org.joda.time.format.DateTimeFormat.forPattern
import java.lang.Integer.parseInt
import java.util.*

const val PREFIX_JSON_UNIT_ALIAS = "!{"
const val PREFIX_EXAM = "\${exam."
const val PREFIX_VAR = "\${#"
const val POSTFIX = '}'

fun resolveJson(body: String, eval: Evaluator): String = eval.resolveJson(body)

fun resolveXml(body: String, eval: Evaluator): String = eval.resolveXml(body)

fun Evaluator.resolveJson(body: String): String = resolve(body, "json", this)

fun Evaluator.resolveXml(body: String): String = resolve(body, "xml", this)

private fun resolve(body: String, type: String, eval: Evaluator): String {
    return resolveAliases(
            type,
            resolveExamCommands(
                    resolveVars(body, eval)
            )
    )
}

private fun resolveVars(body: String, eval: Evaluator): String {
    var result = body
    while (result.contains(PREFIX_VAR)) {
        val original = result
        val v = extractVarFrom(original, PREFIX_VAR)
        result = original.replace(PREFIX_VAR + v + POSTFIX, getObject(eval, v).toString())
    }
    return result
}

private fun getObject(eval: Evaluator, value: String): Any? {
    fun extractAsDate(eval: Evaluator, v: String) = fromDateFields(getObject(eval, v) as Date)
    return if (value.contains(":")) {
        val (v, p) = value.split(":".toRegex(), 2)
        forPattern(p).print(extractAsDate(eval, v))
    } else if (value.contains("+")) {
        val (v, p) = value.split("\\+".toRegex(), 2)
        extractAsDate(eval, v).plus(parsePeriod(p)).toDate()
    } else if (value.contains("-")) {
        val (v, p) = value.split("-".toRegex(), 2)
        extractAsDate(eval, v).minus(parsePeriod(p)).toDate()
    } else {
        eval.getVariable("#$value") ?: eval.evaluate(if (value.contains(".")) "#$value" else value)
    }
}

private fun resolveExamCommands(body: String): String {
    var b = body
    while (b.contains(PREFIX_EXAM)) {
        var original = b
        val v = extractVarFrom(original, PREFIX_EXAM)
        original = original.replace(PREFIX_EXAM + v + POSTFIX, resolveDate(v)!!.toString())
        b = original
    }
    return b
}

private fun resolveAliases(type: String, body: String): String {
    var b = body
    while (b.contains(PREFIX_JSON_UNIT_ALIAS)) {
        val original = b
        val alias = extractFromAlias(original)
        b = original.replace(PREFIX_JSON_UNIT_ALIAS + alias + POSTFIX, toPlaceholder(alias, type))
    }
    return b
}

private fun toPlaceholder(alias: String, type: String): String {
    return when (alias.toLowerCase()) {
        "any-string", "string", "str" -> "\${$type-unit.any-string}"
        "any-number", "number", "num" -> "\${$type-unit.any-number}"
        "any-boolean", "boolean", "bool" -> "\${$type-unit.any-boolean}"
        "regexp", "regex" -> "\${$type-unit.regex}"
        "ignored", "ignore" -> "\${$type-unit.ignore}"
        else -> String.format("\${$type-unit.matches:%s}%s", *alias.split(" ").toTypedArray() as Array<*>)
    }
}

private fun constants(v: String): Any? {
    return when {
        v.startsWith("date(") -> {
            val date = v.substring("date(".length, v.indexOf(")"))
            LocalDateTime.parse(date, forPattern("dd.MM.yyyy")).toDate()
        }
        v.startsWith("now+") -> now().plus(parsePeriod(v.substring(4))).toDate()
        v.startsWith("now-") -> now().minus(parsePeriod(v.substring(4))).toDate()
        else -> when (v) {
            "yesterday" -> now().minusDays(1).toDate()
            "today", "now" -> now().toDate()
            "tomorrow" -> now().plusDays(1).toDate()
            else -> null
        }
    }
}

private fun parsePeriod(v: String): Period {
    return v.substring(1, v.indexOf("]")).split(",")
            .map {
                val (p1, p2) = it.trim().split(" ")
                if (p1.isNum())
                    periodBy(parseInt(p1), p2)
                else
                    periodBy(parseInt(p2), p1)
            }
            .fold(Period.ZERO) { a, n -> a.plus(n) }
}

fun periodBy(value: Int, type: String): BaseSingleFieldPeriod {
    return when (type) {
        "d", "day", "days" -> Days.days(value)
        "M", "month", "months" -> Months.months(value)
        "y", "year", "years" -> Years.years(value)
        "h", "hour", "hours" -> Hours.hours(value)
        "m", "min", "minute", "minutes" -> Minutes.minutes(value)
        "s", "sec", "second", "seconds" -> Seconds.seconds(value)
        else -> throw UnsupportedOperationException("Unsupported period type $type")
    }
}

private fun String.isNum(): Boolean {
    return try {
        parseInt(this)
        true
    } catch (e: NumberFormatException) {
        false
    }
}

fun Evaluator.resolveToObj(placeholder: String?): Any? = when {
    placeholder == null -> null
    placeholder.startsWith("'") && placeholder.endsWith("'") -> placeholder.substring(1, placeholder.lastIndex)
    placeholder.startsWith(PREFIX_VAR) -> getObject(this, extractVarFrom(placeholder, PREFIX_VAR))
    placeholder.startsWith(PREFIX_EXAM) -> resolveDate(extractVarFrom(placeholder, PREFIX_EXAM))
    placeholder.isRange() -> placeholder.toRange()
    else -> placeholder
}

fun resolveToObj(placeholder: String?, evaluator: Evaluator): Any? = evaluator.resolveToObj(placeholder)

fun String.isRange() = this.matches("^[0-9]+[.]{2}[0-9]+$".toRegex())

fun String.toRange(): IntProgression {
    if (this.isRange()) {
        val (start, end) = this.split("[.]{2}".toRegex()).map(String::toInt)
        return IntProgression.fromClosedRange(start, end, end.compareTo(start))
    }
    throw IllegalArgumentException("Couldn't parse range from string $this")
}

private fun extractVarFrom(placeholder: String, namespace: String): String {
    val s = placeholder.substring(placeholder.indexOf(namespace))
    return s.substring(namespace.length, s.indexOf(POSTFIX))
}

private fun extractFromAlias(placeholder: String): String {
    val s = placeholder.substring(placeholder.indexOf(PREFIX_JSON_UNIT_ALIAS))
    return s.substring(2, s.indexOf(POSTFIX))
}

private fun resolveDate(v: String): Any? {
    return if (v.contains(":")) getDateFromPattern(v) else constants(v)
}

private fun getDateFromPattern(value: String): String {
    val (v, p) = value.split(":".toRegex(), 2)
    return forPattern(p).print(fromDateFields((constants(v) as Date?)!!))
}